d2/src/main/java/com/linkedin/d2/balancer/clients/DegraderTrackerClient.java

```diff
@@ -15,173 +15,32 @@
-     *
-     * Pros and Cons:
-     * Deadlocks will be gone since there will be only one DegraderImpl.
-     * However, now it becomes harder to have different configurations for different partitions.
-     */
-    int mapSize = partitionDataMap.size();
-    Map<Integer, PartitionState>partitionStates = new HashMap<Integer, PartitionState>(mapSize * 2);
-    config.setName("TrackerClient Degrader: " + uri);
-    DegraderImpl degrader = new DegraderImpl(config);
-    DegraderControl degraderControl = new DegraderControl(degrader);
-    for (Map.Entry<Integer, PartitionData> entry : partitionDataMap.entrySet())
-    {
-      int partitionId = entry.getKey();
-      PartitionState partitionState = new PartitionState(entry.getValue(), degrader, degraderControl);
-      partitionStates.put(partitionId, partitionState);
-    }
-    _partitionStates = Collections.unmodifiableMap(partitionStates);
-  }
-
-  public Degrader getDegrader(int partitionId)
-  {
-    return getPartitionState(partitionId).getDegrader();
-  }
-
-  public DegraderControl getDegraderControl(int partitionId)
-  {
-
-    return getPartitionState(partitionId).getDegraderControl();
-  }
-
-  private PartitionState getPartitionState(int partitionId)
-  {
-    PartitionState partitionState = _partitionStates.get(partitionId);
-    if (partitionState == null)
-    {
-      String msg = "PartitionState does not exist for partitionId: " + partitionId + ". The current states are " + _partitionStates;
-      throw new IllegalStateException(msg);
-    }
-    return partitionState;
-  }
-
-  private class PartitionState
-  {
-    private final Degrader _degrader;
-    private final DegraderControl _degraderControl;
-    private final PartitionData _partitionData;
-
-    PartitionState(PartitionData partitionData, Degrader degrader, DegraderControl degraderControl)
-    {
-      _partitionData = partitionData;
-      _degrader = degrader;
-      _degraderControl = degraderControl;
-    }
-
-    Degrader getDegrader()
-    {
-      return _degrader;
-    }
-
-    DegraderControl getDegraderControl()
-    {
-      return _degraderControl;
-    }
-
-    PartitionData getPartitionData()
-    {
-      return _partitionData;
-    }
-
-    @Override
-    public String toString()
-    {
-      StringBuilder sb = new StringBuilder();
-      sb.append("{_partitionData = ");
-      sb.append(_partitionData);
-      sb.append(", _degrader = " + _degrader);
-      sb.append(", degraderMinCallCount = " + _degraderControl.getMinCallCount());
-      sb.append("}");
-      return sb.toString();
-    }
-  }
+  /**
+   * @param partitionId Partition ID.
+   * @return Degrader corresponding to the given partition.
+   */
+  Degrader getDegrader(int partitionId);
+
+  /**
+   * @param partitionId Partition ID.
+   * @return DegraderControl corresponding to the given partition.
+   */
+  DegraderControl getDegraderControl(int partitionId);
```

Comment:
Any plan to get rid of DegraderControl? Adding a separate class just for immutability seems really unnecessary.

Reply:
No, this is old code that we  are not touching. All improvements will be made in the new relative strategy.

Reply:
make sense.